### ISSUE TITLE
`suspicious_map` ignores side effects

### ISSUE THREAD

**Issue by zvavybir** [rust-lang/rust-clippy#7767]
`suspicious_map` ignores possible side effects of the map.  That was already mentioned [#5253](https://github.com/rust-lang/rust-clippy/issues/5253) and fixed, but only if it mutates something and not other kinds of side effects.

Lint name:
suspicious_map

I tried this code:

```rust
use std::sync::mpsc::Sender;

fn send_iter_and_get_len<I: Iterator>(iter: I, tx: Sender<I::Item>) -> usize
{
    iter.map(|x| tx.send(x)).count()
}
```

I expected to see this happen: No clippy warnings.

Instead, this happened: I got the following warning and wrong suggestions:
```
warning: this call to `map()` won't have an effect on the call to `count()`
 --> src/main.rs:5:5
  |
5 |     iter.map(|x| tx.send(x)).count()
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(clippy::suspicious_map)]` on by default
  = help: make sure you did not confuse `map` with `filter` or `for_each`
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#suspicious_map
```

### Meta

**Rust version (`rustc -Vv`):**
```
rustc 1.57.0-nightly (9dbb26efe 2021-10-03)
binary: rustc
commit-hash: 9dbb26efe8a886f7bba94c2b996c9535ce07f917
commit-date: 2021-10-03
host: x86_64-unknown-linux-gnu
release: 1.57.0-nightly
LLVM version: 13.0.0
```

**Comment by xFrednet** [comment]
I sadly don't see an easy way to fix this without creating a bunch of false negatives. I would also consider this lint to be correct here. Is there a reason why you used `map`? If you're only trying to interact/inspect each element, then it might be better to use [`inspect`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.inspect) instead. It might be good to reference `inspect()` in the lint description. :upside_down_face:

**Comment by zvavybir** [comment]
I already thought that I might be missing a more appropriate method, but I figured it's still an issue since it should be mentioned.  So I agree that it can be just an documentation edit.  I'll open a pull request, or?

**Comment by xFrednet** [comment]
It would be awesome if you could create a PR. :upside_down_face:  The description can be found here:

https://github.com/rust-lang/rust-clippy/blob/abe551ecb7f0f6223f3d343a591eb34d0b332708/clippy_lints/src/methods/mod.rs#L1281-L1297
